[DirectX] Add DXIL validation of `llvm.loop` metadata

inbelic opened the pull request:
This pr adds the equivalent validation of `llvm.loop` metadata that is [done in DXC](https://github.com/microsoft/DirectXShaderCompiler/blob/8f21027f2ad5dcfa63a275cbd278691f2c8fad33/lib/DxilValidation/DxilValidation.cpp#L3010).

This is done as follows:
- Add `llvm.loop` to the metadata allow-list in `DXILTranslateMetadata`
- Iterate through all `llvm.loop` metadata nodes and strip all incompatible ones
- Raise an error for ill-formed nodes that are compatible with DXIL

Resolves: https://github.com/llvm/llvm-project/issues/137387

inbelic committed:
[DirectX] add a DXILValidateMetadata pass

inbelic committed:
nfc: factor out validations in translate metadata

inbelic committed:
[DirectX] add "llvm.loop" to metadata whitelist

inbelic committed:
[DirectX] add DXIL validation of "llvm.loop" metadata

inbelic commented on llvm/test/CodeGen/DirectX/metadata-stripping.ll:
```diff
@@ -24,9 +24,8 @@ _Z4mainDv3_j.exit:                                ; preds = %for.body.i, %entry
-; CHECK: [[RANGEMD]] = !{i32 1, i32 5}
-; CHECK-NOT: !{!"llvm.loop.mustprogress"}
+; CHECK-DAG: [[RANGEMD]] = !{i32 1, i32 5}
+; CHECK-DAG: !{!"llvm.loop.mustprogress"}
```
This check is changed as `llvm.loop` is now added to the whitelist so it will be allowed through the `dxil-translate-metadata` pass, and then an error would be rasied in `dxil-validate-metadata`

hekota commented on llvm/docs/DirectX/DXILArchitecture.rst:
```diff
@@ -122,6 +122,8 @@ Each of these passes has a defined responsibility:
    to be inserted.
 #. DXILTranslateMetadata transforms and emits all recognized DXIL Metadata.
+#. DXILValidateMetadata validates all emitted DXIL metadata structures
+   conform to DXIL validation.
```
```suggestion
#. DXILValidateMetadata validates that all emitted DXIL metadata structures
   conform to DXIL validation.
```

hekota commented on llvm/lib/Target/DirectX/DXILValidateMetadata.h:
```diff
@@ -0,0 +1,24 @@
+//===- DXILValidateMetadata.h - Pass to emit DXIL metadata -----*- C++ -*-===//
```
```suggestion
//===- DXILValidateMetadata.h - Pass to validate DXIL metadata -*- C++ -*-===//
```

hekota commented on llvm/lib/Target/DirectX/DXILValidateMetadata.cpp:
```diff
@@ -0,0 +1,216 @@
+namespace {
+
+/// A simple Wrapper DiagnosticInfo that generates Module-level diagnostic
+/// for the ValidateMetadata pass
```
```suggestion
/// A simple wrapper of DiagnosticInfo that generates module-level diagnostic
/// for the DXILValidateMetadata pass.
```

hekota commented on llvm/lib/Target/DirectX/DXILValidateMetadata.cpp:
```diff
@@ -0,0 +1,216 @@
+    for (BasicBlock &BB : F)
+      for (Instruction &I : BB) {
+        if (MDNode *LoopMD = I.getMetadata(MDLoopKind))
+          validateLoopMetadata(M, LoopMD);
```
`!llvm.loop` can be only on `br`, right? If so, it would be good to do a `dyn_cast<BranchInst>` first instead of calling `getMetadata` on every instruction in the module. `dyn_cast` is probably a lot faster than `getMetadata`.

hekota reviewed (commented):
Looks good overall, just a few comments and suggestions.

inbelic committed:
review: alias std::array

inbelic committed:
review: fix up comments

inbelic committed:
review: only validate loop metadata on a branch inst

inbelic committed:
review: fix up testcase

inbelic committed:
review: add additional checks for full/disable

inbelic committed:
self-review: merge DXILValidate into DXILTranslate

it seemed like we could create this nice level of abstraction, however,
this will just cause us to write duplicate logic for iterating through
the metadata to first transform and then validate. So we may as well
transform in place

This problem arises as we want to strip certain llvm.loop metadata and
validate on the other types

inbelic committed:
self-review: strip all unrecognized metadata instead of errors

hekota commented on llvm/test/CodeGen/DirectX/metadata-stripping.ll:
```diff
@@ -24,9 +24,9 @@ _Z4mainDv3_j.exit:                                ; preds = %for.body.i, %entry
-; CHECK-NOT: !{!"llvm.loop.mustprogress"}
+; CHECK-DAG: [[RANGEMD]] = !{i32 1, i32 5}
+; CHECK-DAG: [[LOOPMD]] = distinct !{[[LOOPMD]], [[PROGRESS:![0-9]+]]}
+; CHECK-DAG: {!"llvm.loop.mustprogress"}
```
```suggestion
; CHECK-DAG: [[LOOPMD]] = distinct !{[[LOOPMD]], [[PROGRESS:![0-9]+]]}
; CHECK-DAG: [[PROGRESS]] = {!"llvm.loop.mustprogress"}
```
Since you named `PROGRESS` you might as well use it. Or do not name it at all.

inbelic committed:
self-review: missed clean-up

hekota commented on llvm/test/CodeGen/DirectX/Metadata/loop-md-stripped.ll:
```diff
@@ -0,0 +1,58 @@
+
+; CHECK: ![[#LOOP_MD:]] = distinct !{![[#LOOP_MD]]}
+
+!1 = !{!1, i32 0} ; not a metadata node
```
```suggestion
!1 = !{!1, i32 0} ; second operand is not a metadata node
```

hekota replied:
I wonder if the whole node should be stripped instead of just removing the unexpected operand. What does DXC do?

hekota commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -314,16 +327,106 @@ static void translateBranchMetadata(Module &M, Instruction *BBTerminatorInst) {
+        M, "\"llvm.loop.unroll.disable\" and \"llvm.loop.unroll.disable\" "
+           "must be provided as a single operand");
+
+  return true;
```
It would be cleaner if `reportError` just reported the error and did not return `true`/`false`. It should be up to the `isLoopMDCompatible` routine to do that.

hekota commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -314,16 +327,106 @@ static void translateBranchMetadata(Module &M, Instruction *BBTerminatorInst) {
+  ArrayRef<MDOperand> Operands = BaseMD->operands();
+  for (auto &Op : Operands.drop_front())
+    if (isLoopMDCompatible(M, Op.get()))
+      CompatibleOperands.push_back(Op.get());
```
Please add a test case with unrecognized loop metadata hint that gets stripped.

hekota replied:
Oh, I see you have one, nevermind!

inbelic committed:
review: comment clean up

inbelic committed:
review: explicitly return validity

inbelic committed:
self-review: correct naming

hekota reviewed (approved):
LGTM!

bob80905 commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -389,31 +499,23 @@ static void translateGlobalMetadata(Module &M, DXILResourceMap &DRM,
-    M.getContext().diagnose(DiagnosticInfoTranslateMD(
-        M, "Non-library shader: One and only one entry expected"));
-  }
+  } else if (1 < MMDI.EntryPropertyVec.size())
```
nit: this conditional change is unnecessary.

bob80905 commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -314,25 +327,122 @@ static void translateBranchMetadata(Module &M, Instruction *BBTerminatorInst) {
+    }
+  } else if (HintMD->getNumOperands() != 1) {
+    reportLoopError(
+        M, "\"llvm.loop.unroll.disable\" and \"llvm.loop.unroll.disable\" "
```
You probably meant to mention ... .unroll.full as the second thing

bob80905 commented on llvm/test/CodeGen/DirectX/metadata-stripping.ll:
```diff
@@ -24,9 +24,9 @@ _Z4mainDv3_j.exit:                                ; preds = %for.body.i, %entry
-; CHECK: [[RANGEMD]] = !{i32 1, i32 5}
-; CHECK-NOT: !{!"llvm.loop.mustprogress"}
-!0 = distinct !{!0, !1}
+; CHECK-DAG: [[RANGEMD]] = !{i32 1, i32 5}
```
Shouldn't we keep the CHECK-NOT?

inbelic committed:
review: improve error messages + handling

inbelic committed:
review: clarify comments

inbelic committed:
review: correct test-cases

bob80905 commented on llvm/test/CodeGen/DirectX/metadata-stripping.ll:
```diff
@@ -24,9 +24,10 @@ _Z4mainDv3_j.exit:                                ; preds = %for.body.i, %entry
 ; These next check lines check that only the range metadata remains
 ; No more metadata should be necessary, the rest (the current 0 and 1)
 ; should be removed.
-; CHECK-NOT: !{!"llvm.loop.mustprogress"}
```
nit: I don't see why there are differences here.  Maybe some extra whitespace?

bob80905 reviewed (approved):
LGTM just a nit

inbelic committed:
review: fix formatting

inbelic committed:
review: fix formatting
